File: src/InheritanceBasedPolymorphism/Animal.cs
using System;

namespace InheritanceBasedPolymorphism
{
    public abstract class Animal
    {
        public abstract string MakeNoise();
        public abstract string Fly();
    }

    public class Dog : Animal
    {
        public override string MakeNoise() => "Bark";

        public override string Fly() => throw new NotImplementedException();
    }

    public class Cat : Animal
    {
        public override string MakeNoise() => "Meow";

        public override string Fly() => throw new NotImplementedException();
    }

    public class Goose : Animal
    {
        public override string MakeNoise() => "Honk";

        public override string Fly() => "The goose flies away!";
    }
}

File: src/HaskellPolymorphism/Animal.hs
{-# OPTIONS -Wall #-}

module Animal
  ( Animal(..)
  , Dog(..)
  , Cat(..)
  , Goose(..)
  , MakeNoise(..)
  , CanFly(..)
  ) where

class MakeNoise a where
  makeNoise :: a -> String

class CanFly a where
  fly :: a -> String

data Animal
  = Dog' Dog
  | Cat' Cat
  | Goose' Goose
  deriving (Show)

data Dog =
  Dog
  deriving (Show)

instance MakeNoise Dog where
  makeNoise _ = "Bark"

data Cat =
  Cat
  deriving (Show)

instance MakeNoise Cat where
  makeNoise _ = "Meow"

data Goose =
  Goose
  deriving (Show)

instance MakeNoise Goose where
  makeNoise _ = "Honk"

instance CanFly Goose where
  fly _ = "The goose flies away!"
